### PR TITLE
feat(naiveBayesClassifier): Adds support for single field classification with binary data sets using naive bayes

### PR DESCRIPTION
**Pull request by RohanSreerama5**
Awesome! So you want to use Flux for ML? Let's dive in! 

Notes: 
- First pass at supporting Naive Bayes Classification in Flux. 
- Currently supports classification based on a single field
- Supports binary data sets 

Coming very soon: 
- Support for multiple fields 
- Encapsulation of training code into its own function so that training and testing can be performed independently 
- Ability to save training models in InfluxDB bucket 
- Blog post and ReadMe outlining demo and inspiration 

In the works for the future: 
- Implementation for common density functions in order to support non-binary datasets 
- Ability to save training models in SQL table 

original repo: https://github.com/RohanSreerama5/Naive-Bayes-Classifier-Flux

### CODE REVIEW

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/script.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
```
Comment (nathanielc) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/script.py:1]: Can you make this script part of the README for the package? That way its easier for people to discover?

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/zoo_data.csv [hunk 2]
```diff
@@ -0,0 +1,102 @@
+Animal_name,hair,feathers,eggs,milk,airborne,aquatic,predator,toothed,backbone,breathes,venomous,fins,legs,tail,domestic,catsize,type
```
Comment (nathanielc) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/zoo_data.csv:1]: Would it be possible to create a test case out of this data? That way others can see how to use the classifier with pure Flux code? 

To create a test case create a `bayes_test.flux` file (call it whatever you think is best it just needs to end in `_test.flux`) and call the `naiveBayes` function from within that test code. You can embed the csv directly as a string in that file. See the other _test.flux files there are lots of examples to follow.

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux [hunk 3]
```diff
@@ -0,0 +1,105 @@
+//Final working code as of August 11, 2020
+//Currently supports single field classification and binary data sets 
+
+import "system"
```
Comment (nathanielc) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:4]: I think you need to run `make generate` so that this code gets converted into some Go code to correctly include it into the Flux build. See the Flux README for instructions on how to do that.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:4]: Based on the guidelines in CONTRIBUTING.md, I ran `go generate ./stdlib` since I'm only contributing .flux files. Is this along the lines of what you're thinking of?

Comment (nathanielc) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:4]: Yes, running the generate command will create some Go files that are effectively the contents of the contributed .flux files. Specifically the Go files contain the parsed AST of the flux files.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:4]: Yeah just noticed that. Currently working on writing some tests!

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+naiveBayes = (myClass, myBucket, myField, myMeasurement) => {
```
Comment (fchikwekwe) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:6]: Oh this is so cool to see in Flux!! Reminds me of the DS classes that I took. Mad props for working this out in Flux!

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/zoo_data.csv (hunk 2 above)
Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/zoo_data.csv:1]: If you don't know what the `outData` might look like for your test, you can try it out in the Flux repl.

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux (hunk 3 above)
Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:4]: Quick q: So for the tests, I'm trying to use csv.from() and point to the path of my csv file, but my naiveBayes flux code is set up to run a from(bucket: myBucket) to get csv data. Is there a way to write to a bucket in flux and then be able to pass that bucket name as an argument to naiveBayes(...).

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+package naiveBayesClassifier
+
+import "system"
+
+
+
+////////csv.from(csv: naiveBayesClassifier/zoo_data.csv) |> naiveBayes()
+
+//---
```
Comment (fchikwekwe) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:12]: I know this isn't the final version, but just a reminder to go through and remove all the commented out lines of Flux.

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  //|> yield(name: "trainingData")
+
+test_data = tables
+  //from(bucket: myBucket)
+  |> range(start:2020-01-01T00:00:00Z, stop: 2020-01-01T23:00:00Z) //data for 1 day
+  |> filter(fn: (r) => r["_measurement"] == myMeasurement and r["_field"] == myField)
+  |> group()
+  //|> yield(name: "test data")
+
+//data preparation
```
Comment (fchikwekwe) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:31]: I think there are a few ways to make this script reusable so that it can work with more than just the one data set in the future. If you maybe pulled the data prep out of this function and made it its own module and then standardized column naming, then others could use your functions as well. We can discuss this more if you're interested in opening another PR to further modify this. 

For now, I think this is a good guideline for folks who want to use Flux for these types of algorithms in the future.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:32]: Hi Faith,
Just wanted to let you know after some messing around, I got the bayes classifier to finally run make generate successfully. Only things different I've done is install ruby. I'm also not sure if this is related or not, but as a side thing, I cloned down a fresh copy of the flux repo master and when I ran make, it did not make and complained about the '/r' line break character. So I installed a Unix program called dos2unix and ran dos2unix pkg-config.sh which is supposed to convert the line breaks in the text file to Unix format. Once I did this and ran make again, the flux repo ran make successfully. Anyways, my project is now compiling properly on my machine so if you can approve of this, I'll do some code cleanup very soon and we can push this to master! Please get back to me when you get a chance.

**Comment by fchikwekwe** [comment]
Hey @RohanSreerama5! To pass the Semantic PR check, you need to squash all your commits into one using `git` and then change the commit message to fit this convention https://github.com/zeke/semantic-pull-requests. If you have trouble with that, lemme know and I can try to help.

**Comment by fchikwekwe** [comment]
It also looks like your go modules are out of date. If you click on the `Details` link by the failed tests, you'll see that it asks you to run `go mod tidy` on your branch. Try that out and then run `make test` on your machine to see if everything compiles locally before pushing it back up to your remote branch.

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux (hunk 6 above)
Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Yeah, generally things look good now. You do need to pass all the CI checks in order for this to be merged. See the notes I left below.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: I just ran go mod tidy and make test. I think I have another rust issue. This time, I think I need rustup installed, but when I try to install rustup, it panics. I'm searching the web for possible solutions but I'm not too sure where to go with this. Do you have a few mins to spare on Zoom?

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: I might have some time later. What is the panic that you're seeing?

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: It is panicking when I run this to install rustup. 
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `22`,
 right: `4`', src/libstd/sys/unix/thread.rs:179:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `22`,
 right: `4`', src/libstd/sys/unix/thread.rs:179:21
stack backtrace:
   0:     0x7f7bf9edaffd - backtrace::backtrace::libunwind::trace::h812748238d609e46
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x7f7bf9edaffd - backtrace::backtrace::trace_unsynchronized::h7c97e818aebf09c8
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x7f7bf9edaffd - std::sys_common::backtrace::_print_fmt::h60d914263b0ccd71
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x7f7bf9edaffd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf78227137afc7565
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x7f7bf9b6524c - core::fmt::write::h543cdf60775f89bf
                               at src/libcore/fmt/mod.rs:1069
   5:     0x7f7bf9eda904 - std::io::Write::write_fmt::h0c7f3ce24c679426
                               at src/libstd/io/mod.rs:1504
   6:     0x7f7bf9eda105 - std::sys_common::backtrace::_print::h80e55e24be231368
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x7f7bf9eda105 - std::sys_common::backtrace::print::h3b197b9c1261c865
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x7f7bf9eda105 - std::panicking::default_hook::{{closure}}::ha6c807149ce20f8f
                               at src/libstd/panicking.rs:198
   9:     0x7f7bf9ed98f4 - std::panicking::default_hook::he49a9c12e358cc45
                               at src/libstd/panicking.rs:218
  10:     0x7f7bf9ed93b6 - std::panicking::rust_panic_with_hook::h93f74f5ef2f71f31
                               at src/libstd/panicking.rs:515
  11:     0x7f7bf9ed9198 - rust_begin_unwind
                               at src/libstd/panicking.rs:419
  12:     0x7f7bf9ed9140 - std::panicking::begin_panic_fmt::hfa6ef29ba81f400e
                               at src/libstd/panicking.rs:373
  13:     0x7f7bf9e11671 - <rustup::diskio::threaded::Threaded as rustup::diskio::Executor>::join::habf6ee901a47a8a4
  14:     0x7f7bf9e10c78 - core::ptr::drop_in_place::h0b1b0dcb4b07c267
  15:     0x7f7bf9d76c40 - core::ptr::drop_in_place::h09572fbec04e583f
  16:     0x7f7bf9e2842e - rustup::dist::component::package::unpack_without_first_dir::h99f4b0c2cae349ff
  17:     0x7f7bf9de559f - rustup::dist::manifestation::Manifestation::update::hacf67fad3b44f03a
  18:     0x7f7bf9dd563c - rustup::dist::dist::update_from_dist_::h1462627f9c58c494
  19:     0x7f7bf9dd227a - rustup::install::InstallMethod::install::hbb6430b2dfaf5792
  20:     0x7f7bf9dd0d21 - rustup::toolchain::DistributableToolchain::install_from_dist::h7b9bf9317c9aa2c5
  21:     0x7f7bf9ec5aa2 - rustup::cli::self_update::install::h8ca0f61a773526a9
  22:     0x7f7bf9ecbd3d - rustup::cli::setup_mode::main::h4d335e1871435d70
  23:     0x7f7bf9adf42a - rustup_init::main::h22da971c634937a2
  24:     0x7f7bf9ef3e83 - std::rt::lang_start_internal::{{closure}}::{{closure}}::h4ed4ab1fb893cc93
                               at src/libstd/rt.rs:52
  25:     0x7f7bf9ef3e83 - std::sys_common::backtrace::__rust_begin_short_backtrace::h1f01c818c00c4f70
                               at src/libstd/sys_common/backtrace.rs:130
  26:     0x7f7bf9ae053f - main
  27:     0x7f7bf96270b3 - __libc_start_main
  28:     0x7f7bf9adb029 - <unknown>
thread panicked while panicking. aborting.
Illegal instruction (core dumped)

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Basically, make test is complaining that I don't have a default Rust toolchain set

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: I think you mentioned that you're now working on a Windows device. You may still be using the Mac installation method.  I think that the Windows method should be accessible using this webpage: https://rustup.rs

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Yeah, that website is actually where I copied the curl command to install rust.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: if you look at the website, it says not to use the curl command for Windows. There an exe file that you download instead.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: O actually, I'm on Windows Subsystem for Linux. So it would make sense to use that. Sorry, should've been more clear.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Yeah, I'm not an expert on running Windows w/ Linux, but I would try the Windows instructions.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Or if that can't be done, then the Windows/Linux clash might be part of the issue here. I think I'd be out of my depth if that's the case.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: I've been using it for a while for school stuff too and its basically just a Linux machine that I'm running. So I'm not entirely sure its an OS problem. I'll keep looking into it though -thx!

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: hi. Its the end of my day here, but this guide should set you on the right course. https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec

First, you'll have a merge conflict with `stdlib/test_packages.go`. When you try to rebase it will ask you to fix that. 

Then, start following the guide. You can do: 
- `git rebase -i HEAD~9`, then you should see a list of all your commits with the `pick` command there. If you're not set up with interactive rebase you'll be in vim and you can press `i` to start editing. 
- Replace all the `pick` with `squash` except the first one. This will squash all your subsequent commits go into the first one. Replace the first `pick` with `reword` while one that same screen. 
- YOu'll have another prompt to reword the commit. Follow the semantic commit message convention by writing a commit message that completes the sentence: "This commit will...."

Give that a try. If you're having trouble ping me here early if your day and I'll see if there's time on my calendar to pair.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Example of how I'd reword your first commit: 
`feat(stdlib/contrib): add support for single field classification using naive bayes`

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Hi, I was able to squash the commits via rebase locally. But I'm a bit unsure about how to make this change show up on the Git remote branch

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Hey rohan. I'm not at my computer right now, but if you type the command  `git push --force`, it should override the remote commits and replace them with your local ones. I'd double and triple check that your local branch is correct first by checking `git status` and then making sure all your changes are up to date.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Yeah. Now it's complaining about my PR title, which is odd because I did use the word 'feat' in it. Also, before I forget, I wanted to mention something about the code. On lines 43 and 81 of naiveBayesClassifier.flux, there exists some code that is specific to the Animal dataset that I had used. Specifically on line 43 where I do `r.airborne`, last I spoke with Anais, she told me that in Flux there isn't a way to turn that into a variable and do something like `r.{myField}`.  I think for the time being, we can push this up to master and address these nits soon after if this sounds alright to you.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: yeah, I added a merge commit because you still had a merge conflict on this branch that you hadn't fixed yet. I fixed it for you, but I can't update the commit message since its not my branch. It will just have to fail that check for now.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Yeah there's a lot of stuff in this code that is specific to your example and may not work modularly with other people's datasets. I think its okay to revisit it when you have time since its in a contrib folder.

Comment (fchikwekwe) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: @RohanSreerama5 i just got back. I've got a few minutes if you'd like to resolve the test failure.

Comment (RohanSreerama5) [reply stdlib/contrib/RohanSreerama5/naiveBayesClassifier/naiveBayesClassifier.flux:23]: Sure I'm game. I'll send you a zoom link via email

**Comment by fchikwekwe** [comment]
@RohanSreerama5 I'm just gonna merge this into a feature branch in the flux repo so that I can pull it down, fix the test failures and then merge it into master. I'll add you as a reviewer once I get the tests passing so that you can make sure that any changes reflect your plan for this branch. Hopefully get that done sometime today.

**Comment by fchikwekwe** [comment]
@RohanSreerama5 the new PR is #3191. Please take a look today and leave a comment if everything looks okay.

File: stdlib/contrib/RohanSreerama5/naiveBayesClassifier/README.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+3) You will have to make edits to script.py in the following areas: 
+
+- Set `mydata` equal to the file path of your dataset 
+- Fill in your values for `token`, `bucket`, and `organization`. (If you are using my sample zoo dataset you can ignore the rest of this step) 
+- Ensure that `periods` is set equal to the value you get from running the `data.size` cell. 
+- Set `dataframe_measurement_name` to be the name of your dataset. 
+- Note which fields you want to use as actual fields you classify on and which you would like to use for Class. Recall we predict `P(Class | field)`.
+- Anything you choose to be a class must be listed inside of `data_frame_tag_columns` and everything else defaults to a `field`. Finally, run the script. You've just written your dataset to an InfluxDB bucket. 
+
+![](images/pythonScript.png)
```
Comment (fchikwekwe) [inline stdlib/contrib/RohanSreerama5/naiveBayesClassifier/README.md:25]: @RohanSreerama5 Nathaniel pointed out that we don't typically added images to READMEs with data that can be pasted in as text. Would you mind removing the images and either adding this data as text or linking to the existing csvs in this project that contain the data?
